File: Packages/BioMedLit/Sources/BioMedLit/Transparency/Analysis/FundingAnalyzer.swift
// BMLibrarian Lite - Biomedical Literature Research Tool
// Copyright (C) 2024-2025 Dr Horst Herb
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/>.

import Foundation

/// Pure functions for analyzing study funding sources.
///
/// All functions are stateless and can be safely called from any context.
/// This module provides classification of funders as industry or non-industry,
/// determination of overall sponsor types, and parsing of funder data from
/// various sources (CrossRef, PubMed).
///
/// Usage:
/// ```swift
/// let classification = FundingAnalyzer.classifyFunder(name: "Pfizer Inc.", doi: "10.13039/100004319")
/// // Returns: (isIndustry: true, confidence: 1.0)
///
/// let sponsorType = FundingAnalyzer.determineSponsorType(from: funders)
/// // Returns: .industry, .government, .academic, .mixed, or .unknown
/// ```
public enum FundingAnalyzer {

    // MARK: - Confidence Thresholds

    /// Confidence level for known industry funder DOI matches.
    private static let knownFunderConfidence: Double = 1.0

    /// Confidence level for government pattern matches.
    private static let governmentPatternConfidence: Double = 0.85

    /// Confidence level for academic pattern matches.
    private static let academicPatternConfidence: Double = 0.80

    /// Confidence level for corporate suffix matches.
    private static let corporateSuffixConfidence: Double = 0.75

    /// Confidence level for industry keyword matches.
    private static let industryKeywordConfidence: Double = 0.70

    /// Default confidence level when classification is uncertain.
    private static let unknownConfidence: Double = 0.30

    /// Number of primary industry keywords to check before falling back to broader patterns.
    private static let primaryIndustryKeywordCount: Int = 6

    // MARK: - Funder Classification

    /// Classify a single funder as industry or non-industry.
    ///
    /// Uses a multi-layer approach to determine if a funder is an industry entity:
    /// 1. Check known industry funder DOIs (highest confidence)
    /// 2. Check government/academic name patterns (takes precedence over corporate indicators)
    /// 3. Check corporate name indicators (legal entity suffixes)
    /// 4. Check broader industry keywords
    ///
    /// - Parameters:
    ///   - name: Funder name (required).
    ///   - doi: CrossRef Funder Registry DOI (optional, e.g., "10.13039/100004319").
    /// - Returns: Tuple of (isIndustry, confidence 0.0-1.0) where confidence indicates
    ///   reliability of the classification.
    public static func classifyFunder(name: String, doi: String? = nil) -> (isIndustry: Bool, confidence: Double) {
        // Layer 1: Check known industry funder DOIs (highest confidence)
        if let doi = doi, KnownIndustryFunders.isIndustryFunder(doi) {
            return (true, knownFunderConfidence)
        }

        let nameLower = name.lowercased()

        // Layer 2: Check government/academic patterns first
        // These take precedence even if corporate suffixes are present
        if RegexHelper.anyMatch(patterns: IndustryPatterns.governmentPatterns, in: nameLower) {
            return (false, governmentPatternConfidence)
        }

        if RegexHelper.anyMatch(patterns: IndustryPatterns.academicPatterns, in: nameLower) {
            return (false, academicPatternConfidence)
        }

        // Layer 3: Check corporate indicators
        if RegexHelper.anyMatch(patterns: IndustryPatterns.corporateSuffixes, in: nameLower) {
            return (true, corporateSuffixConfidence)
        }

        // Layer 4: Check broader industry keywords (first few are more specific)
        let primaryPatterns = Array(IndustryPatterns.industryKeywords.prefix(primaryIndustryKeywordCount))
        if RegexHelper.anyMatch(patterns: primaryPatterns, in: nameLower) {
            return (true, industryKeywordConfidence)
        }

        // Unknown - low confidence
        return (false, unknownConfidence)
    }

    /// Create a FunderInfo from raw funder data.
    ///
    /// Combines the funder classification with the provided metadata to create
    /// a complete FunderInfo object suitable for use in transparency analysis.
    ///
    /// - Parameters:
    ///   - name: Funder name (required).
    ///   - doi: CrossRef Funder Registry DOI (optional).
    ///   - awardNumbers: Grant/award numbers associated with this funder.
    /// - Returns: FunderInfo with industry classification populated.
    public static func createFunderInfo(
        name: String,
        doi: String? = nil,
        awardNumbers: [String] = []
    ) -> FunderInfo {
        let (isIndustry, confidence) = classifyFunder(name: name, doi: doi)
        return FunderInfo(
            name: name,
            funderDOI: doi,
            awardNumbers: awardNumbers,
            isIndustry: isIndustry,
            confidence: confidence
        )
    }

    // MARK: - Sponsor Type Determination

    /// Determine overall sponsor type from list of funders.
    ///
    /// Analyzes the collection of funders to determine the primary sponsor type:
    /// - `.industry` if only industry funders are present
    /// - `.mixed` if both industry and non-industry funders are present
    /// - `.government` if only government funders are present (no industry)
    /// - `.academic` if only academic funders are present (no industry, no government)
    /// - `.nonprofit` if non-industry funders don't match government/academic patterns
    /// - `.unknown` if no funders are provided
    ///
    /// - Parameter funders: List of analyzed funders.
    /// - Returns: Overall sponsor type classification.
    public static func determineSponsorType(from funders: [FunderInfo]) -> SponsorType {
        guard !funders.isEmpty else {
            return .unknown
        }

        let industryFunders = funders.filter { $0.isIndustry }
        let nonIndustryFunders = funders.filter { !$0.isIndustry }

        // Pure industry funding
        if !industryFunders.isEmpty && nonIndustryFunders.isEmpty {
            return .industry
        }

        // Mixed funding
        if !industryFunders.isEmpty && !nonIndustryFunders.isEmpty {
            return .mixed
        }

        // No industry funding - check if government or academic
        let hasGovernment = nonIndustryFunders.contains { funder in
            RegexHelper.anyMatch(
                patterns: IndustryPatterns.governmentPatterns,
                in: funder.name.lowercased()
            )
        }

        if hasGovernment {
            return .government
        }

        let hasAcademic = nonIndustryFunders.contains { funder in
            RegexHelper.anyMatch(
                patterns: IndustryPatterns.academicPatterns,
                in: funder.name.lowercased()
            )
        }

        return hasAcademic ? .academic : .nonprofit
    }

    /// Determine industry funding status from funders.
    ///
    /// Returns whether any industry funding was detected and the maximum
    /// confidence level among industry funders.
    ///
    /// - Parameter funders: List of analyzed funders.
    /// - Returns: Tuple of (detected, maxConfidence) where detected is true
    ///   if any industry funder was found.
    public static func industryFundingStatus(from funders: [FunderInfo]) -> (detected: Bool, confidence: Double) {
        let industryFunders = funders.filter { $0.isIndustry }

        guard !industryFunders.isEmpty else {
            return (false, 0.0)
        }

        let maxConfidence = industryFunders.map(\.confidence).max() ?? 0.0
        return (true, maxConfidence)
    }

    // MARK: - CrossRef Data Parsing

    /// Parse funders from CrossRef work response.
    ///
    /// Extracts funder information from the CrossRef API "funder" array,
    /// creating classified FunderInfo objects for each entry.
    ///
    /// Expected dictionary structure:
    /// ```json
    /// {
    ///   "name": "Pfizer Inc.",
    ///   "DOI": "10.13039/100004319",
    ///   "award": ["R01-123456"]
    /// }
    /// ```
    ///
    /// - Parameter crossRefFunders: Array of funder dictionaries from CrossRef API.
    /// - Returns: List of FunderInfo objects with classifications.
    public static func parseCrossRefFunders(_ crossRefFunders: [[String: Any]]?) -> [FunderInfo] {
        guard let crossRefFunders = crossRefFunders else { return [] }

        return crossRefFunders.compactMap { funderDict -> FunderInfo? in
            guard let name = funderDict["name"] as? String else { return nil }

            let doi = funderDict["DOI"] as? String
            let awards = funderDict["award"] as? [String] ?? []

            return createFunderInfo(name: name, doi: doi, awardNumbers: awards)
        }
    }

    /// Parse funders from PubMed grant data.
    ///
    /// Extracts funder information from PubMed's grant list format,
    /// creating classified FunderInfo objects for each entry.
    ///
    /// Expected dictionary structure:
    /// ```json
    /// {
    ///   "agency": "National Institutes of Health",
    ///   "grant_id": "R01-CA123456"
    /// }
    /// ```
    ///
    /// - Parameter grants: Array of grant dictionaries from PubMed.
    /// - Returns: List of FunderInfo objects with classifications.
    public static func parsePubMedGrants(_ grants: [[String: String?]]?) -> [FunderInfo] {
        guard let grants = grants else { return [] }

        return grants.compactMap { grant -> FunderInfo? in
            // Unwrap the double optional: grant["agency"] returns String??
            guard let agencyOpt = grant["agency"],
                  let agency = agencyOpt,
                  !agency.isEmpty else {
                return nil
            }

            // Similarly unwrap grant_id (optional value from optional key)
            let grantId: String? = grant["grant_id"].flatMap { $0 }
            let awardNumbers = grantId.map { [$0] } ?? []

            return createFunderInfo(name: agency, awardNumbers: awardNumbers)
        }
    }

    // MARK: - Funder Deduplication

    /// Merge funders from multiple sources, removing duplicates.
    ///
    /// Funders are considered duplicates if their names match (case-insensitive).
    /// When duplicates are found, the one with higher confidence is kept,
    /// ensuring the most reliable classification is retained.
    ///
    /// - Parameter funderLists: Variable number of funder arrays from different sources.
    /// - Returns: Deduplicated list of funders with highest confidence retained.
    public static func mergeFunders(_ funderLists: [FunderInfo]...) -> [FunderInfo] {
        var fundersByName: [String: FunderInfo] = [:]

        for list in funderLists {
            for funder in list {
                let key = funder.name.lowercased()

                if let existing = fundersByName[key] {
                    // Keep the one with higher confidence
                    if funder.confidence > existing.confidence {
                        fundersByName[key] = funder
                    }
                } else {
                    fundersByName[key] = funder
                }
            }
        }

        return Array(fundersByName.values)
    }

    // MARK: - Trial Sponsor Analysis

    /// Update sponsor type based on clinical trial sponsor class.
    ///
    /// Incorporates sponsor information from ClinicalTrials.gov to refine
    /// the sponsor type classification. If the trial indicates industry
    /// sponsorship but the current type doesn't, upgrades to mixed funding.
    ///
    /// - Parameters:
    ///   - currentType: Current sponsor type based on funding analysis.
    ///   - trialSponsorClass: Sponsor class from ClinicalTrials.gov
    ///     (e.g., "INDUSTRY", "NIH", "OTHER").
    /// - Returns: Updated sponsor type that incorporates trial sponsor information.
    public static func updateSponsorType(
        _ currentType: SponsorType,
        withTrialSponsorClass trialSponsorClass: String?
    ) -> SponsorType {
        guard let sponsorClass = trialSponsorClass?.uppercased() else {
            return currentType
        }

        let isTrialIndustry = sponsorClass == "INDUSTRY"

        switch currentType {
        case .unknown:
            return isTrialIndustry ? .industry : currentType
        case .government, .academic, .nonprofit:
            return isTrialIndustry ? .mixed : currentType
        case .industry, .mixed:
            return currentType
        }
    }

    // MARK: - Summary

    /// Generate a human-readable summary of funding analysis.
    ///
    /// - Parameters:
    ///   - funders: List of analyzed funders.
    ///   - sponsorType: Overall sponsor type classification.
    /// - Returns: Summary string suitable for display in UI.
    public static func formatSummary(funders: [FunderInfo], sponsorType: SponsorType) -> String {
        guard !funders.isEmpty else {
            return "No funding information available"
        }

        let industryCount = funders.filter { $0.isIndustry }.count
        let totalCount = funders.count

        switch sponsorType {
        case .industry:
            return "Industry-funded (\(industryCount) funder\(industryCount == 1 ? "" : "s"))"
        case .mixed:
            return "Mixed funding (\(industryCount) industry, \(totalCount - industryCount) non-industry)"
        case .government:
            return "Government-funded (\(totalCount) funder\(totalCount == 1 ? "" : "s"))"
        case .academic:
            return "Academic-funded (\(totalCount) funder\(totalCount == 1 ? "" : "s"))"
        case .nonprofit:
            return "Non-profit funded (\(totalCount) funder\(totalCount == 1 ? "" : "s"))"
        case .unknown:
            return "Funding source unknown"
        }
    }
}

File: android/MedicalFactChecker/app/src/main/java/com/bmlibrarian/factchecker/domain/transparency/FundingAnalyzer.kt
package com.bmlibrarian.factchecker.domain.transparency

import kotlinx.serialization.json.JsonObject

/**
 * How [FundingAnalyzer.classifyFunder] classified one funder.
 *
 * @property isIndustry Whether the funder is classified as industry.
 * @property confidence Confidence (0.0-1.0), which names the layer that decided (see [FundingAnalyzer]).
 */
data class FunderClassification(val isIndustry: Boolean, val confidence: Double)

/**
 * Whether a set of funders includes industry, and how sure that is.
 *
 * @property detected True if any funder is industry.
 * @property confidence The highest confidence among the industry funders; 0.0 when none.
 */
data class IndustryFundingStatus(val detected: Boolean, val confidence: Double)

/**
 * Pure functions classifying study funders and sponsor types.
 *
 * Ported from the Swift `FundingAnalyzer` (BioMedLit). The confidence each
 * classification layer reports is pinned by the shared contract
 * `sponsor_patterns.json` (`confidences`), and the industry boundary is measured
 * against `funder_names.json`; both are asserted by this platform's tests.
 */
object FundingAnalyzer {

    /** Confidence of a known industry funder DOI match. */
    private const val KNOWN_FUNDER_CONFIDENCE: Double = 1.0

    /** Confidence of a government pattern match. */
    private const val GOVERNMENT_PATTERN_CONFIDENCE: Double = 0.85

    /** Confidence of an academic pattern match. */
    private const val ACADEMIC_PATTERN_CONFIDENCE: Double = 0.80

    /** Confidence of a funder-name stem or whole-word match. */
    private const val INDUSTRY_NAME_CONFIDENCE: Double = 0.75

    /** Confidence when nothing matched. */
    private const val UNKNOWN_CONFIDENCE: Double = 0.30

    /** The sponsor class ClinicalTrials.gov reports for an industry sponsor. */
    private const val INDUSTRY_SPONSOR_CLASS: String = "INDUSTRY"

    /**
     * Classify a single funder as industry or not.
     *
     * Layers, first match wins:
     *  1. a known industry funder DOI (highest confidence);
     *  2. government, then academic, name patterns — these outrank corporate
     *     suffixes ("Department of Veterans Affairs" is not a company);
     *  3. the calibrated funder-name stems and whole words.
     *
     * Layer 3, with the brand list (#394), scores precision 0.958 / recall 0.657
     * on the shared labelled corpus, against 0.455 / 0.167 for the substring
     * matcher it replaced.
     *
     * @param name Funder name.
     * @param doi CrossRef Funder Registry DOI, if known (e.g. "10.13039/100004319").
     * @return The classification and its confidence.
     */
    fun classifyFunder(name: String, doi: String? = null): FunderClassification {
        if (doi != null && KnownIndustryFunders.isIndustryFunder(doi)) {
            return FunderClassification(isIndustry = true, confidence = KNOWN_FUNDER_CONFIDENCE)
        }

        val nameLower = name.lowercase()

        if (TransparencyRegex.anyMatch(IndustryPatterns.governmentPatterns, nameLower)) {
            return FunderClassification(isIndustry = false, confidence = GOVERNMENT_PATTERN_CONFIDENCE)
        }
        if (TransparencyRegex.anyMatch(IndustryPatterns.academicPatterns, nameLower)) {
            return FunderClassification(isIndustry = false, confidence = ACADEMIC_PATTERN_CONFIDENCE)
        }
        if (matchesIndustryName(nameLower)) {
            return FunderClassification(isIndustry = true, confidence = INDUSTRY_NAME_CONFIDENCE)
        }
        return FunderClassification(isIndustry = false, confidence = UNKNOWN_CONFIDENCE)
    }

    /**
     * Whether a lowercased funder name carries an industry stem or whole-word marker.
     *
     * The single predicate behind both funder sources (CrossRef `funder[].name`
     * and PubMed `<Grant><Agency>`). Deliberately not applied to COI prose; see
     * [IndustryPatterns.industryKeywords].
     */
    private fun matchesIndustryName(nameLower: String): Boolean =
        IndustryPatterns.funderNameStems.any { nameLower.contains(it) } ||
            TransparencyRegex.anyMatch(IndustryPatterns.funderNameWords, nameLower) ||
            matchesIndustryBrand(nameLower)

    /**
     * Whether a lowercased funder name names a known company, and not its foundation (#394).
     *
     * @param nameLower The funder name, already lowercased.
     * @return True if an [IndustryPatterns.funderBrandPatterns] entry matches and no
     *   [IndustryPatterns.foundationMarkerPatterns] entry does.
     */
    internal fun matchesIndustryBrand(nameLower: String): Boolean =
        TransparencyRegex.anyMatch(IndustryPatterns.funderBrandPatterns, nameLower) &&
            !TransparencyRegex.anyMatch(IndustryPatterns.foundationMarkerPatterns, nameLower)

    /**
     * Create a classified [FunderInfo] from raw funder data.
     *
     * @param name Funder name.
     * @param doi CrossRef Funder Registry DOI, if known.
     * @param awardNumbers Grant or award numbers.
     * @return The funder with its industry classification.
     */
    fun createFunderInfo(
        name: String,
        doi: String? = null,
        awardNumbers: List<String> = emptyList(),
    ): FunderInfo {
        val classification = classifyFunder(name, doi)
        return FunderInfo(
            name = name,
            funderDOI = doi,
            awardNumbers = awardNumbers,
            isIndustry = classification.isIndustry,
            confidence = classification.confidence,
        )
    }

    /**
     * The overall sponsor type of a set of funders.
     *
     * [SponsorType.INDUSTRY] if every funder is industry; [SponsorType.MIXED] if
     * some are; otherwise [SponsorType.GOVERNMENT] if any funder's name matches a
     * government pattern, else [SponsorType.ACADEMIC] if any matches an academic
     * one, else [SponsorType.NONPROFIT]. [SponsorType.UNKNOWN] when there are no funders.
     *
     * @param funders Classified funders.
     * @return The sponsor type.
     */
    fun determineSponsorType(funders: List<FunderInfo>): SponsorType {
        if (funders.isEmpty()) return SponsorType.UNKNOWN

        val industryFunders = funders.filter { it.isIndustry }
        val nonIndustryFunders = funders.filter { !it.isIndustry }

        if (industryFunders.isNotEmpty() && nonIndustryFunders.isEmpty()) return SponsorType.INDUSTRY
        if (industryFunders.isNotEmpty()) return SponsorType.MIXED

        val hasGovernment = nonIndustryFunders.any {
            TransparencyRegex.anyMatch(IndustryPatterns.governmentPatterns, it.name.lowercase())
        }
        if (hasGovernment) return SponsorType.GOVERNMENT

        val hasAcademic = nonIndustryFunders.any {
            TransparencyRegex.anyMatch(IndustryPatterns.academicPatterns, it.name.lowercase())
        }
        return if (hasAcademic) SponsorType.ACADEMIC else SponsorType.NONPROFIT
    }

    /**
     * Whether any funder is industry, with the highest industry confidence.
     *
     * @param funders Classified funders.
     * @return The status; `(false, 0.0)` when there is no industry funder.
     */
    fun industryFundingStatus(funders: List<FunderInfo>): IndustryFundingStatus {
        val industryFunders = funders.filter { it.isIndustry }
        if (industryFunders.isEmpty()) return IndustryFundingStatus(detected = false, confidence = 0.0)
        return IndustryFundingStatus(detected = true, confidence = industryFunders.maxOf { it.confidence })
    }

    /**
     * Classify the funders of a CrossRef work's `funder` array.
     *
     * Each entry is read as Swift reads it: `name` must be a string (entries
     * without one are skipped), `DOI` is used only when a string, and `award` is
     * used only when every element is a string.
     *
     * @param crossRefFunders The `funder` entries, already known to be JSON objects; null for none.
     * @return The classified funders, in source order.
     */
    fun parseCrossRefFunders(crossRefFunders: List<JsonObject>?): List<FunderInfo> {
        if (crossRefFunders == null) return emptyList()
        return crossRefFunders.mapNotNull { funder ->
            val name = JsonCasts.string(funder["name"]) ?: return@mapNotNull null
            val doi = JsonCasts.string(funder["DOI"])
            val awards = JsonCasts.stringList(funder["award"]) ?: emptyList()
            createFunderInfo(name = name, doi = doi, awardNumbers = awards)
        }
    }

    /**
     * Classify funders from PubMed grant entries (`agency` / `grant_id`).
     *
     * @param grants Grant maps; entries whose agency is null or empty are skipped. Null for none.
     * @return The classified funders, in source order.
     */
    fun parsePubMedGrants(grants: List<Map<String, String?>>?): List<FunderInfo> {
        if (grants == null) return emptyList()
        return grants.mapNotNull { grant ->
            val agency = grant["agency"]
            if (agency.isNullOrEmpty()) return@mapNotNull null
            val awardNumbers = grant["grant_id"]?.let { listOf(it) } ?: emptyList()
            createFunderInfo(name = agency, awardNumbers = awardNumbers)
        }
    }

    /**
     * Merge funder lists from several sources, dropping duplicates.
     *
     * Funders are duplicates when their names match case-insensitively; the one
     * with the higher confidence is kept. Unlike Swift, which returns a
     * dictionary's values in no defined order, the result here is in the order
     * each name was first seen.
     *
     * @param funderLists The lists to merge.
     * @return One funder per name.
     */
    fun mergeFunders(vararg funderLists: List<FunderInfo>): List<FunderInfo> {
        val fundersByName = LinkedHashMap<String, FunderInfo>()
        for (list in funderLists) {
            for (funder in list) {
                val key = funder.name.lowercase()
                val existing = fundersByName[key]
                if (existing == null || funder.confidence > existing.confidence) {
                    fundersByName[key] = funder
                }
            }
        }
        return fundersByName.values.toList()
    }

    /**
     * Refine a sponsor type with a ClinicalTrials.gov sponsor class.
     *
     * An industry trial sponsor turns an unknown sponsor type into industry, and
     * a government/academic/non-profit one into mixed.
     *
     * @param currentType The sponsor type from the funding analysis.
     * @param trialSponsorClass The trial's sponsor class (e.g. "INDUSTRY", "NIH"), or null.
     * @return The refined sponsor type.
     */
    fun updateSponsorType(currentType: SponsorType, trialSponsorClass: String?): SponsorType {
        val sponsorClass = trialSponsorClass?.uppercase() ?: return currentType
        val isTrialIndustry = sponsorClass == INDUSTRY_SPONSOR_CLASS
        return when (currentType) {
            SponsorType.UNKNOWN -> if (isTrialIndustry) SponsorType.INDUSTRY else currentType
            SponsorType.GOVERNMENT, SponsorType.ACADEMIC, SponsorType.NONPROFIT ->
                if (isTrialIndustry) SponsorType.MIXED else currentType
            SponsorType.INDUSTRY, SponsorType.MIXED -> currentType
        }
    }

    /**
     * A one-line summary of a funding analysis for display.
     *
     * @param funders Classified funders.
     * @param sponsorType Their sponsor type.
     * @return The summary.
     */
    fun formatSummary(funders: List<FunderInfo>, sponsorType: SponsorType): String {
        if (funders.isEmpty()) return "No funding information available"

        val industryCount = funders.count { it.isIndustry }
        val totalCount = funders.size

        return when (sponsorType) {
            SponsorType.INDUSTRY -> "Industry-funded ($industryCount funder${plural(industryCount)})"
            SponsorType.MIXED ->
                "Mixed funding ($industryCount industry, ${totalCount - industryCount} non-industry)"
            SponsorType.GOVERNMENT -> "Government-funded ($totalCount funder${plural(totalCount)})"
            SponsorType.ACADEMIC -> "Academic-funded ($totalCount funder${plural(totalCount)})"
            SponsorType.NONPROFIT -> "Non-profit funded ($totalCount funder${plural(totalCount)})"
            SponsorType.UNKNOWN -> "Funding source unknown"
        }
    }

    /** The plural suffix for [count] items: "" for one, "s" otherwise. */
    private fun plural(count: Int): String = if (count == 1) "" else "s"
}
